code.js

```diff
@@ -1,3 +1,71 @@
+
 function tsp_hk(distance_matrix) {
-    return -1;
+    // Sizes of 0 or 1 have no distance
+    let size = distance_matrix.length;
+    if (size == 0 || size == 1) { return 0; }
+    
+
+    let bestMin = Infinity; // Initialize min
+
+    // Make Every City the Starting Node
+    for (let i = 0; i < size; i++) {
+        let dictionaryCache = {};
+        let cities = [];
+
+        // Fill array with the all cities
+        for (let j = 0; j < size; j++) { cities.push(j); }
+
+        let currentMin = heldKarp(i, cities, distance_matrix, dictionaryCache);
+
+        // Replaces current min with new min
+        if (currentMin < bestMin) { bestMin = currentMin; }
+    }
+    
+    return bestMin;
 }
+
+
+// Based on the psuedocode in the README
+function heldKarp(city, cities, matrix, cache) {
+    let size = cities.length;
+
+    // Check if it exists already in cache
+    let key = JSON.stringify([city, cities]);
```

Comment:
What if the cities are in a different order, but the same set?

Reply:
The way that the cities array is generated (in the tsp_hk() function) would mean that it is always in sorted order. Or am I mistaken?

Reply:
You could convince me that this is true :)

Reply:
![Untitled video - Made with Clipchamp](https://github.com/user-attachments/assets/8c0527dc-4e3d-441a-ab2e-7d85ec2e18bf)